File: carddemo/app/services/interest_calculator.py
"""Interest calculator — from CBACT04C.cbl paragraph 1300-COMPUTE-INTEREST.

COBOL formula (line 464-465):
  COMPUTE WS-MONTHLY-INT = (TRAN-CAT-BAL * DIS-INT-RATE) / 1200

This preserves the exact business logic:
  monthly_interest = (tran_cat_balance * interest_rate) / 1200

The 1200 divisor converts annual rate (e.g., 18.00%) to monthly:
  18.00 / 1200 = 0.015 = 1.5% monthly
"""

from decimal import ROUND_HALF_UP, Decimal

from sqlalchemy.orm import Session

from app.models.account import Account
from app.models.card_xref import CardXref
from app.models.disclosure_group import DisclosureGroup
from app.models.tran_category_balance import TranCategoryBalance


def calculate_monthly_interest(balance: Decimal, rate: Decimal) -> Decimal:
    """Exact replica of CBACT04C 1300-COMPUTE-INTEREST.

    COMPUTE WS-MONTHLY-INT = (TRAN-CAT-BAL * DIS-INT-RATE) / 1200
    """
    if rate == 0:
        return Decimal("0.00")
    return (balance * rate / Decimal("1200")).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)


def run_interest_calculation(db: Session) -> dict:
    """Full interest calculation batch — replaces CBACT04C main loop.

    Reads all transaction category balance records sequentially,
    looks up disclosure group for interest rate,
    computes interest, and updates account balances.
    """
    processed = 0
    total_interest = Decimal("0.00")
    last_acct_id = None
    acct_total_interest = Decimal("0.00")

    tcat_records = db.query(TranCategoryBalance).order_by(TranCategoryBalance.acct_id).all()

    for tcat in tcat_records:
        # When account changes, update the previous account (from 1050-UPDATE-ACCOUNT)
        if last_acct_id is not None and tcat.acct_id != last_acct_id:
            _update_account_balance(db, last_acct_id, acct_total_interest)
            acct_total_interest = Decimal("0.00")

        last_acct_id = tcat.acct_id

        # Look up account to get group_id (from 1100-GET-ACCT-DATA)
        account = db.query(Account).filter_by(acct_id=tcat.acct_id).first()
        if not account:
            continue

        # Look up interest rate from disclosure group (from 1200-GET-INTEREST-RATE)
        rate = _get_interest_rate(db, account.group_id, tcat.type_cd, tcat.cat_cd)

        if rate != Decimal("0"):
            monthly_int = calculate_monthly_interest(Decimal(str(tcat.balance)), rate)
            acct_total_interest += monthly_int
            total_interest += monthly_int
            processed += 1

    # Update the last account
    if last_acct_id is not None:
        _update_account_balance(db, last_acct_id, acct_total_interest)

    db.commit()
    return {"processed": processed, "total_interest": str(total_interest)}


def _get_interest_rate(db: Session, group_id: str, type_cd: str, cat_cd: int) -> Decimal:
    """Look up interest rate — from CBACT04C 1200-GET-INTEREST-RATE.

    Falls back to 'DEFAULT' group if specific group not found (status '23').
    """
    disc = db.query(DisclosureGroup).filter_by(group_id=group_id, tran_type_cd=type_cd, tran_cat_cd=cat_cd).first()
    if disc:
        return Decimal(str(disc.interest_rate))

    # Fallback to DEFAULT group (from 1200-A-GET-DEFAULT-INT-RATE)
    disc = db.query(DisclosureGroup).filter_by(group_id="DEFAULT", tran_type_cd=type_cd, tran_cat_cd=cat_cd).first()
    if disc:
        return Decimal(str(disc.interest_rate))

    return Decimal("0")


def _update_account_balance(db: Session, acct_id: int, total_interest: Decimal) -> None:
    """Update account balance with interest — from CBACT04C 1050-UPDATE-ACCOUNT.

    ADD WS-TOTAL-INT TO ACCT-CURR-BAL
    MOVE 0 TO ACCT-CURR-CYC-CREDIT
    MOVE 0 TO ACCT-CURR-CYC-DEBIT
    """
    account = db.query(Account).filter_by(acct_id=acct_id).first()
    if account:
        account.curr_bal = Decimal(str(account.curr_bal)) + total_interest
        account.curr_cyc_credit = Decimal("0")
        account.curr_cyc_debit = Decimal("0")


def get_xref_card_num(db: Session, acct_id: int) -> str:
    """Get card number from cross-reference — from CBACT04C 1110-GET-XREF-DATA."""
    xref = db.query(CardXref).filter_by(acct_id=acct_id).first()
    return xref.card_num if xref else ""

File: java-migration/src/main/java/com/carddemo/batch/job/InterestCalculator.java
package com.carddemo.batch.job;

import com.carddemo.batch.model.AccountRecord;
import com.carddemo.batch.model.CardXrefRecord;
import com.carddemo.batch.model.DisclosureGroupRecord;
import com.carddemo.batch.model.TranCatBalRecord;
import com.carddemo.batch.model.TransactionRecord;
import com.carddemo.batch.util.DateFormatUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.BufferedWriter;
import java.io.IOException;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Java equivalent of COBOL program CBACT04C.
 * Interest calculator batch program that:
 *   1. Reads transaction category balance records sequentially
 *   2. For each account, looks up account data, cross-reference, and disclosure group
 *   3. Computes monthly interest = (balance * rate) / 1200
 *   4. Writes interest transaction records to output file
 *   5. Updates account balances
 */
public class InterestCalculator {

    private static final Logger log = LoggerFactory.getLogger(InterestCalculator.class);
    private static final String DEFAULT_GROUP = "DEFAULT";

    private final List<TranCatBalRecord> tranCatBalRecords;
    private final Map<Long, AccountRecord> accountMap;
    private final Map<Long, CardXrefRecord> xrefByAcctId;
    private final Map<String, DisclosureGroupRecord> disclosureGroupMap;
    private final Path transactionOutputFile;
    private final String parmDate;

    private int recordCount;
    private int tranIdSuffix;

    public InterestCalculator(
            List<TranCatBalRecord> tranCatBalRecords,
            Map<Long, AccountRecord> accountMap,
            Map<Long, CardXrefRecord> xrefByAcctId,
            Map<String, DisclosureGroupRecord> disclosureGroupMap,
            Path transactionOutputFile,
            String parmDate) {
        this.tranCatBalRecords = tranCatBalRecords;
        this.accountMap = accountMap;
        this.xrefByAcctId = xrefByAcctId;
        this.disclosureGroupMap = disclosureGroupMap;
        this.transactionOutputFile = transactionOutputFile;
        this.parmDate = parmDate;
        this.recordCount = 0;
        this.tranIdSuffix = 0;
    }

    /**
     * Execute the interest calculation batch job.
     *
     * @return list of generated interest transaction records
     */
    public List<TransactionRecord> execute() throws IOException {
        log.info("START OF EXECUTION OF PROGRAM CBACT04C (InterestCalculator)");
        List<TransactionRecord> generatedTransactions = new ArrayList<>();

        try (BufferedWriter tranWriter = Files.newBufferedWriter(transactionOutputFile)) {
            long lastAcctNum = -1;
            boolean firstTime = true;
            BigDecimal totalInt = BigDecimal.ZERO;
            AccountRecord currentAccount = null;

            for (TranCatBalRecord catBal : tranCatBalRecords) {
                recordCount++;
                log.debug("{}", catBal);

                if (catBal.getTrancatAcctId() != lastAcctNum) {
                    if (!firstTime && currentAccount != null) {
                        updateAccount(currentAccount, totalInt);
                    } else {
                        firstTime = false;
                    }

                    totalInt = BigDecimal.ZERO;
                    lastAcctNum = catBal.getTrancatAcctId();

                    currentAccount = accountMap.get(catBal.getTrancatAcctId());
                    if (currentAccount == null) {
                        log.warn("ACCOUNT NOT FOUND: {}", catBal.getTrancatAcctId());
                        continue;
                    }

                    CardXrefRecord xref = xrefByAcctId.get(catBal.getTrancatAcctId());
                    if (xref == null) {
                        log.warn("XREF NOT FOUND FOR ACCOUNT: {}", catBal.getTrancatAcctId());
                    }
                }

                if (currentAccount == null) {
                    continue;
                }

                DisclosureGroupRecord discGroup = getInterestRate(
                        currentAccount.getAcctGroupId(),
                        catBal.getTrancatTypeCd(),
                        catBal.getTrancatCd());

                if (discGroup != null && discGroup.getDisIntRate().compareTo(BigDecimal.ZERO) != 0) {
                    BigDecimal monthlyInt = computeInterest(catBal.getTranCatBal(), discGroup.getDisIntRate());
                    totalInt = totalInt.add(monthlyInt);

                    CardXrefRecord xref = xrefByAcctId.get(catBal.getTrancatAcctId());
                    String cardNum = xref != null ? xref.getXrefCardNum() : "";

                    TransactionRecord txn = createInterestTransaction(
                            currentAccount, monthlyInt, cardNum);
                    generatedTransactions.add(txn);
                    writeTransactionRecord(tranWriter, txn);
                }
            }

            // Update the last account
            if (currentAccount != null) {
                updateAccount(currentAccount, totalInt);
            }
        }

        log.info("END OF EXECUTION OF PROGRAM CBACT04C - processed {} records, generated {} transactions",
                recordCount, generatedTransactions.size());
        return generatedTransactions;
    }

    /**
     * Look up disclosure group interest rate (1200-GET-INTEREST-RATE).
     * Falls back to DEFAULT group if specific group not found.
     */
    DisclosureGroupRecord getInterestRate(String acctGroupId, String tranTypeCd, int tranCatCd) {
        String key = buildDiscGroupKey(acctGroupId, tranTypeCd, tranCatCd);
        DisclosureGroupRecord record = disclosureGroupMap.get(key);

        if (record == null) {
            log.debug("DISCLOSURE GROUP RECORD MISSING - TRY WITH DEFAULT GROUP CODE");
            String defaultKey = buildDiscGroupKey(DEFAULT_GROUP, tranTypeCd, tranCatCd);
            record = disclosureGroupMap.get(defaultKey);
        }

        if (record == null) {
            log.warn("ERROR READING DEFAULT DISCLOSURE GROUP for key: {}", key);
        }

        return record;
    }

    /**
     * Compute monthly interest (1300-COMPUTE-INTEREST).
     * Formula: monthlyInterest = (balance * rate) / 1200
     */
    BigDecimal computeInterest(BigDecimal balance, BigDecimal rate) {
        return balance.multiply(rate)
                .divide(new BigDecimal("1200"), 2, RoundingMode.HALF_UP);
    }

    /**
     * Update account balances after interest calculation (1050-UPDATE-ACCOUNT).
     */
    void updateAccount(AccountRecord account, BigDecimal totalInterest) {
        account.setAcctCurrBal(account.getAcctCurrBal().add(totalInterest));
        account.setAcctCurrCycCredit(BigDecimal.ZERO);
        account.setAcctCurrCycDebit(BigDecimal.ZERO);
        log.debug("Updated account {} balance to {}", account.getAcctId(), account.getAcctCurrBal());
    }

    /**
     * Create an interest transaction record (1300-B-WRITE-TX).
     */
    TransactionRecord createInterestTransaction(AccountRecord account, BigDecimal monthlyInt, String cardNum) {
        tranIdSuffix++;
        TransactionRecord txn = new TransactionRecord();
        txn.setTranId(parmDate + String.format("%06d", tranIdSuffix));
        txn.setTranTypeCd("01");
        txn.setTranCatCd(5);
        txn.setTranSource("System");
        txn.setTranDesc("Int. for a/c " + account.getAcctId());
        txn.setTranAmt(monthlyInt);
        txn.setTranMerchantId(0);
        txn.setTranMerchantName("");
        txn.setTranMerchantCity("");
        txn.setTranMerchantZip("");
        txn.setTranCardNum(cardNum);
        String timestamp = DateFormatUtil.getDb2FormatTimestamp();
        txn.setTranOrigTs(timestamp);
        txn.setTranProcTs(timestamp);
        return txn;
    }

    private void writeTransactionRecord(BufferedWriter writer, TransactionRecord txn) throws IOException {
        writer.write(String.format("%-16s%-2s%04d%-10s%-100s%+012.2f%09d%-50s%-50s%-10s%-16s%-26s%-26s",
                txn.getTranId(),
                txn.getTranTypeCd(),
                txn.getTranCatCd(),
                txn.getTranSource(),
                txn.getTranDesc(),
                txn.getTranAmt(),
                txn.getTranMerchantId(),
                txn.getTranMerchantName(),
                txn.getTranMerchantCity(),
                txn.getTranMerchantZip(),
                txn.getTranCardNum(),
                txn.getTranOrigTs(),
                txn.getTranProcTs()));
        writer.newLine();
    }

    private static String buildDiscGroupKey(String groupId, String typeCd, int catCd) {
        String paddedGroup = groupId == null ? "" : groupId.trim();
        return String.format("%-10s%-2s%04d", paddedGroup, typeCd, catCd);
    }

    public int getRecordCount() {
        return recordCount;
    }
}
